src/System.Linq/tests/ToHashSetTests.cs

```diff
@@ -58,17 +59,184 @@ public void TolerateDuplicates()
+                source => source.Union(Range(20, 50)), // UnionIterator
+                source => source.Select(x => x), // SelectArrayIterator, SelectEnumerableIterator, SelectListIterator, SelectIListIterator
+                source => source.Skip(1).Select(x => x), // SelectListPartitionIterator, SelectIPartitionIterator
+                source => source.Where(x => x < 40), // WhereArrayIterator, WhereEnumerableIterator, WhereListIterator
+                source => source.Where(x => x < 40).Select(x => x) // WhereSelectArrayIterator, WhereSelectEnumerableIterator, WhereSelectListIterator
+            );
+
+            testData.AddSources(GetSources());
+
+            return testData.GetData();
+        }
+
+        private static IEnumerable<object[]> GetGroupingData()
+        {
+            var testData = new WorkAsExpectedData<IGrouping<int, int>>();
+
+            testData.AddDefaultComparers();
+
+            testData.AddActions(
+                source => source.GroupBy(x => x), // GroupedEnumerable<TSource, TKey>, GroupedEnumerable<TSource, TKey, TElement>
+                source => source.GroupBy(x => x, x => x), // GroupedResultEnumerable<TSource, TKey, TResult>, GroupedResultEnumerable<TSource, TKey, TElement, TResult>
+                source => source.ToLookup(i => i) // Lookup
+            );
+
+            testData.AddSources(GetSources());
+
+            return testData.GetData();
+        }
+
+        /// <summary>
+        /// Assert.Equals -> HashSet.SetEquals not work for IGrouping
+        /// </summary>
+        /// <param name="expected">Expected set</param>
+        /// <param name="actual">Actual set</param>
+        private void AssertHashSetGroupingEqual(HashSet<IGrouping<int, int>> expected, HashSet<IGrouping<int, int>> actual)
+        {
+            // The Count should be equal
+            Assert.Equal(expected.Count, actual.Count);
+
+            // Get the group by key and check as IEnumerable
+            foreach (var expectedGroup in expected)
+            {
+                var group = actual.FirstOrDefault(x => x.Key == expectedGroup.Key);
+
+                Assert.NotNull(group);
+
+                Assert.Equal(expectedGroup.AsEnumerable(), group.AsEnumerable());
+            }
+        }
+        
+        private static IEnumerable<IEnumerable<int>> GetSources()
+        {
+            // Create empty, distinct and repeated sources
+            var sources = new[]
+            {
+                Array.Empty<int>(),
+                Range(0, 50),
+                Range(0, 50).Concat(Range(20, 50))
+            };
+
+            // Some methods have specific handling for Arrays, Lists and IEnumerables
+            foreach (var source in sources)
+            {
+                yield return source.ToArray();
+                yield return new List<int>(source);
+                yield return new SortedSet<int>(source);
```

Comment:
From what I can see, all of these classes implement `ICollection<T>`, so for iterators that are specialized for `ICollection<T>` this will only cover the specialized case. You may want to copy some of the logic from [here](https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/coding-style.md) so you're more likely to cover all the branches. (Note: If you choose to copy the logic, you can delete the big comment and the `GetType()` check because covariant arrays are not being used in the test data here.)